Post excerpt block - add color, fontSize, lineHeight, and text alignment.

Addison-Stavlo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
As part of #21087 to add feature parity to FSE blocks.  This adds settings (and some necessary restructuring/styles to support) text alignment and the experimental flags for color, fontSize, and lineHeight.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Tested on local docker env.

Verify settings applied and saved in the editor for colors, font size, line height, and text alignment are appropriately reflected in the editor and on the front-end.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
add text align

Addison-Stavlo committed:
add wrapper and line-height inherit

Addison-Stavlo committed:
php format

noahtallen reviewed (commented):
For some reason, I'm not seeing the excerpt show up. I tried putting it in a `singular` template with a page selected. It also seems like alignment options don't show up in this case either...

<img width="636" alt="Screen Shot 2020-07-14 at 9 11 56 PM" src="https://user-images.githubusercontent.com/6265975/87502666-f4fca680-c616-11ea-9de9-d8f6cc74c3dd.png">

But font size and color work!

By the way, I think the failing php test is due to phpcs issues in the PHP file

Copons commented:
+1 for not being able to test this.

I had the very same issue for Post Date in #23931, and solved it by explicitly using the `postId` from the context in `useEntityProps`.

E.g.
```json
// block.json

{ "context": [ "postType", "postId" ] }
```

```js
// edit.js

function Edit( { context: { postType, postId } ) {
  const [ excerpt, setExcerpt ] = useEntityProp( 'postType', postType, 'excerpt', postId );
  // ...
}
```

In #23931 this also had the advantage of simplifying the block, as it doesn't need the wrapping export anymore.

I've also noticed a few things that we could improve on this block, but that aren't strictly related to this PR.
I'll leave some inline comments mostly for me to remember them, but they are in no way blockers! 🙂

Addison-Stavlo committed:
get this working in site editor?

Addison-Stavlo commented on packages/block-library/src/post-excerpt/edit.js:
```diff
@@ -128,15 +135,18 @@ export default function PostExcerptEdit( {
-	if ( ! useEntityId( 'postType', 'post' ) ) {
-		return __( 'Post Excerpt' );
-	}
+	// This check doesn't work in site editor? always false.
+	// if ( ! useEntityId( 'postType', 'post' ) ) {
+	// 	return __( 'Post Excerpt' );
+	// }
```
This early return is always getting triggered in the site editor.  Im not sure what else to use as a replacement.

Copons replied:
Just for reference, for the Post Date (#23931), I've chosen to completely remove the wrapping function, letting the edit function handle the existence of the post attribute.

E.g. (very simplified)

Before
```js
function PostDateEditor( context ) {
  // Get the date with useEntityProp and the post type and ID contained in the context
  return <div>{ date ? date : 'No date' }</div>;
}

export default function PostDateEdit( { context } ) {
  if ( ! context ) { return 'Post date'; }
  return PostDateEditor( context );
}
```

After
```js
export default function PostDateEdit( { context } ) {
  // Get the date with useEntityProp and the post type and ID contained in the context
  return <div>{ date ? date : 'No date' }</div>;
}
```

Addison-Stavlo reviewed (commented):
Updated so it 'works' in the context of selecting a specific post in the site editor's drop-down.

Addison-Stavlo committed:
set initial readMore, get rid of extra parent component

Addison-Stavlo committed:
styles to ensure link color is shown

Addison-Stavlo committed:
no elipsis

Addison-Stavlo committed:
dont initialize for now /shrug

noahtallen commented:
<img width="877" alt="Screen Shot 2020-07-15 at 9 11 46 PM" src="https://user-images.githubusercontent.com/6265975/87625769-ceef0900-c6df-11ea-9e36-cc4d0de7f576.png">

I added post excerpt to the `index` template. when i switched to an individual post, I got that error :/

noahtallen commented:
It looks like it is working for the homepage though!

<img width="1526" alt="Screen Shot 2020-07-15 at 9 13 23 PM" src="https://user-images.githubusercontent.com/6265975/87625872-107fb400-c6e0-11ea-839a-57dc1b42b907.png">

noahtallen reviewed (commented):
The bold/italics buttons can be activated, but don't seem to do anything :/

<img width="761" alt="Screen Shot 2020-07-15 at 9 14 01 PM" src="https://user-images.githubusercontent.com/6265975/87626020-5a689a00-c6e0-11ea-8ebe-cf480fd93b10.png">

Outside of that, it looks like typography and color settings are working correctly! It would be nice to fix the fatal error before merging (if it's reproducible, anyways)

youknowriad commented on packages/block-library/src/post-excerpt/style.scss:
```diff
@@ -4,3 +4,7 @@
 }
+
+.wp-block-post-excerpt.has-link-color a {
+	color: var(--wp--style--color--link) !important;
```
do we need `!important` ?

Addison-Stavlo replied:
Either that or more specificity.  Since there are a handful of nested elements in this block theme supplied generic link style rules overwrite this option when we also have a background.

youknowriad replied:
Do you have an example of rule with higher-specificity and where does it come from? (Just to help with the reasoning)

Addison-Stavlo replied:
Sure thing (maybe there is a better way to deal with this), but currently I am testing the link colors with the Seedlet-Blocks theme (I can't seem to get the option to show up for 2020/2019-blocks).  When a background is applied, the link color selection stops working because of this rule:

![Screen Shot 2020-07-16 at 11 10 56 AM](https://user-images.githubusercontent.com/28742426/87688694-4e6aef80-c755-11ea-8715-7ec1acd177dd.png)

youknowriad replied:
For me, this is an area where the theme should adapt its styles. Ideally, with global styles (and the link color feature is part of global styles), the themes don't provide as much detailed CSS but it's more "managed".

Addison-Stavlo replied:
I was wondering about this as well - Theme styles being too restrictive vs. block styles not defining enough.  Although, I was having trouble finding another theme where I could test link colors anymore (I remember previously testing them on one of the 2020/2019/blocks themes but they don't seem present anymore?).

Since link colors in simple blocks such as paragraphs work, I figured these more complex/nested element blocks should adapt their styles accordingly.  After all `has-background a` does not seem like a very opinionated rule, and makes sense for a theme to supply a default style for links used in background areas.  So perhaps blocks like this need some sort of rule to ensure that if a custom link color is applied, the nested elements are explicitly given a style rule to adhere to?

Copons replied:
I think this is a Seedlet theme bug, though.
That `:not( .has-background-background-color )` increases the specificity enough to override the block's link color.

If you try to remove that `:not()`, you'll see that the block's rule doesn't need the `!important` anymore.

All the same, you could leave the `:not()` on the theme, and add another class to the block's rule to make the color work without `!important`, e.g.
```scss
.wp-block-post-author.has-link-color a,
.wp-block-post-author.has-link-color.has-background a {
	color: var( --wp--style--color--link );
}
```


AFAIK, this happens because `:not( .class )` works as if it added a `.NOT-class` to the selected element.
So you'd have `.wp-block-post-author.has-link-color` (2 "classes" of specificity) selected by the editor styles, and something like `.has-background.NOT-has-background-background-color` (again, 2 classes of specificity) selected by the theme styles.
Since the theme styles are loaded after the editor's, given the same specificity, they prevail.

I'm not entirely sure of the purpose of that `:not( .has-background-background-color )` (defined here: [editor](https://github.com/Automattic/themes/blob/0bff9da40595b98a294ac48b51fca73aa772b028/seedlet/assets/sass/blocks/utilities/_editor.scss#L57-L63), [front end](https://github.com/Automattic/themes/blob/0bff9da40595b98a294ac48b51fca73aa772b028/seedlet/assets/sass/blocks/utilities/_style.scss#L154-L160)), by the way.
It basically forces a text color to all text blocks with a background, except those using the `background` palette color as background color.
I kinda suspect this selector could be revised in the theme, since there's already an override ([editor](https://github.com/Automattic/themes/blob/0bff9da40595b98a294ac48b51fca73aa772b028/seedlet/assets/sass/base/_editor.scss#L34-L38), [front end](https://github.com/Automattic/themes/blob/0bff9da40595b98a294ac48b51fca73aa772b028/seedlet/assets/sass/base/_reset.scss#L96-L100)) for blocks with `p` as root element.

Addison-Stavlo replied:
Very well put, thanks @copons!  Updated 😄

jffng replied:
> themes don't provide as much detailed CSS

Copons commented:
I'm not sure how to test the experimental link color. 🤔 
Should I see a block control? Should I just see the block showing up with the `has-link-color` class? Is it a theme option?
I'm not seeing either (on Seedlet Blocks theme), but I can at least confirm that manually adding the CSS class seems to change the link color (from blue to black/gray, in Seedlet Blocks).

Addison-Stavlo commented:
> The bold/italics buttons can be activated, but don't seem to do anything :/

@noahtallen You need to select text for it to work.  And this block is set up in a very odd way in that the preview of the post's content you pointed out is only a placeholder for the rich text excerpt field.  So if you click on that text and start typing, it will start creating the custom excerpt to save for the post.  This is the actual rich text that can have those styles applied. 🤷‍♀️  

> I'm not sure how to test the experimental link color. 🤔

@Copons it should be available in seedlet blocks.  The 'read more...' section is a link, but you could add a link to the custom excerpt through rich text to test as well:

![link-color](https://user-images.githubusercontent.com/28742426/87684936-d0a4e500-c750-11ea-8258-1356b463fd30.gif)

Addison-Stavlo commented:
> I added post excerpt to the index template. when i switched to an individual post, I got that error :/

Gotcha.  I get this error when I add it to a template in general as postType and postId are undefined (I think triggered by useEntityProp in that case).  

We could return something like what you pointed out before - https://github.com/WordPress/gutenberg/pull/23945#pullrequestreview-448616491 - that was preventing us from seeing the error.  Or change the text to say something like 'No post found.' ?  Or is the error'd block better in this case because it is easy to tell that the block has encountered an error, as opposed to hiding it in some text that looks like a paragraph block?

![Screen Shot 2020-07-16 at 12 11 21 PM](https://user-images.githubusercontent.com/28742426/87695424-7b230500-c75d-11ea-91c1-7109a6024e12.png)

vs

![Screen Shot 2020-07-16 at 12 12 55 PM](https://user-images.githubusercontent.com/28742426/87695659-c210fa80-c75d-11ea-8d20-fa8f70bbf453.png)
This is what was given before by the early return.  It was being triggered everywhere in the site editor, be we could update the condition to have it happen in this case where postType/Id aren't found.  Although I wonder if the explicit error box makes more sense.

Addison-Stavlo committed:
add custom warning if no post found in context

Addison-Stavlo committed:
more simple error fix

Addison-Stavlo committed:
actually we need this warning... otherwise user can edit what cannot be dirtied/saved

noahtallen commented:
> You need to highlight text for it to work 

Ahh that makes sense. I think I expected it to just change the style of the whole thing. (probably an area for design to give feedback, along with how it edits the excerpt inline)

noahtallen commented:
> as it has no corresponding entity to save against.

I think in my brain, it makes the most sense to have it be a placeholder of some sort. Not sure what design would look like in that situation. But I think this would fit with any of the blocks that don't have context available (like post_content block without having a page selected as context). Instead of erroring or warning, i think it should show a placeholder thingy that helps people understand what it would look like when it has the right context. Since it's not really an actual "error" in that the block was inserted in a place where it is allowed to be

Addison-Stavlo commented:
> I think I expected it to just change the style of the whole thing. 

I expected that the first time I tried the block too.  Its definitely a bit funky feeling atm.

Copons commented:
This works well for me, but I'd rather have a different way to [handle front end link colors that doesn't involve `!important`](https://github.com/WordPress/gutenberg/pull/23945#discussion_r455596699).

Addison-Stavlo committed:
get rid of important

ZebulanStanphill commented on packages/block-library/src/post-excerpt/block.json:
```diff
@@ -2,6 +2,9 @@
 	"name": "core/post-excerpt",
 	"category": "design",
 	"attributes": {
+		"align": {
```
In my opinion, this should most likely not be called `align`, but rather `textAlign`. I know `align` is how the Paragraph block does it, but this is confusing since there is also block alignment, which (if using the `supports` flag) uses `align` as its attribute name.

Addison-Stavlo replied:
Agreed, `textAlign` or `alignText` would be preferable.  I saw the `align` support hook is for block alignment as well and noticed the confusion.  I wonder how many other blocks have this conflicted naming convention, but we should start updating them.

Addison-Stavlo replied:
Or would it be better to similarly update the `align` hook to be `blockAlign` or something similar?

If this is more recent than some of the blocks that use `align` for textAlign, then less people should be effected by the change.  That is, Im assuming if we updated the paragraph block's attr name then a user who updates their Gutenberg version will have whatever alignment settings they saved lost?

noahtallen replied:
Is there a compatibility layer for changing a block attribute name? I think there is something for updating a block to a newer version without loosing settings 🤔

Addison-Stavlo replied:
> Is there a compatibility layer for changing a block attribute name? I think there is something for updating a block to a newer version without loosing settings 🤔

That would be good, at the very least for now I will update the FSE blocks that are using `align` for text alignment to use `textAlign`

ZebulanStanphill replied:
There is indeed a way to easily handle (what would otherwise be backward-incompatible) changes in block attribute schemas:

https://developer.wordpress.org/block-editor/developers/block-api/block-deprecation/

A lot of core blocks already make use this, storing their deprecated versions in a `deprecated.js` file.

Addison-Stavlo replied:
Updated for the FSE blocks specifically - https://github.com/WordPress/gutenberg/pull/24077
